### PR TITLE
Add MCP well-known URI auth discovery per MCP Spec

### PR DESCRIPTION
**Pull request by theJC**
Fixes https://github.com/stacklok/toolhive/issues/2526

Implements RFC 9728 Protected Resource Metadata discovery via well-known URIs when WWW-Authenticate header is not present. This completes ToolHive's implementation of the [MCP specification requirement ](https://modelcontextprotocol.io/specification/draft/basic/authorization#protected-resource-metadata-discovery-requirements)that clients MUST support both discovery mechanisms (because any given compliant server MUST implement one of them)

Changes:
- Add tryWellKnownDiscovery() to discover auth via well-known URIs
- Add buildWellKnownURI() to construct RFC 9728 compliant URIs
- Add checkWellKnownURIExists() to validate URI accessibility
- Modify DetectAuthenticationFromServer() for well-known fallback
- Add comprehensive unit tests for all discovery paths
- Add regression tests for empty scope handling contracts

The implementation tries endpoint-specific URIs first, then root-level URIs per MCP spec priority order. This enables authentication with MCP-compliant servers that use well-known URIs per RFC 9728 Section 3 but don't send WWW-Authenticate headers.

### CODE REVIEW

File: pkg/auth/discovery/discovery.go [hunk 1]
```diff
@@ -156,6 +171,93 @@ func detectAuthWithRequest(
...
+			ResourceMetadata: wellKnownURI,
+		}, nil
+	}
+
+	return nil, nil // No well-known metadata found
+}
+
 // ParseWWWAuthenticate parses the WWW-Authenticate header to extract authentication information
 // Supports multiple authentication schemes and complex header formats
 func ParseWWWAuthenticate(header string) (*AuthInfo, error) {
```
Comment (amirejaz) [inline pkg/auth/discovery/discovery.go:275]: Let's also make sure we check the Content-Type header to confirm it's JSON before proceeding:
```
// Check content type
contentType := strings.ToLower(resp.Header.Get("Content-Type"))
if !strings.Contains(contentType, "application/json") {
    return false
}
```

This will help ensure we’re not accepting or later parsing non-JSON responses (like HTML error pages) as JSON (which will throw the error).

Comment (theJC) [reply pkg/auth/discovery/discovery.go:275]: Excellent catch/feedback, thank you.  Addressed as asked.

Comment (theJC) [reply pkg/auth/discovery/discovery.go:275]: Also changed to require a 200 / OK response and to no longer accept 401 as signal that it exists (I missed catching that in my initial self code review of my and claude's work).

Comment (amirejaz) [reply pkg/auth/discovery/discovery.go:275]: Thank you for your contribution and for addressing the feedback so promptly, much appreciated!

**Comment by amirejaz** [comment]
@theJC It looks like the CI linter is failing. Could you please run `task lint` locally to check and fix any linting issues?

File: pkg/auth/discovery/discovery.go [hunk 2]
```diff
@@ -156,6 +171,103 @@ func detectAuthWithRequest(
...
+		baseURL.Path = path.Join(auth.WellKnownOAuthResourcePath, cleanPath)
+	} else {
+		// Root-level: /.well-known/oauth-protected-resource
+		baseURL.Path = auth.WellKnownOAuthResourcePath
+	}
+
+	return baseURL.String()
+}
+
+// checkWellKnownURIExists checks if a well-known URI is accessible and returns valid JSON metadata
```
Comment (amirejaz) [inline pkg/auth/discovery/discovery.go]: nit: Lets remove the `and returns valid JSON metadata` as we are only returning `bool`

File: pkg/auth/discovery/discovery.go [hunk 3]
```diff
@@ -156,6 +171,103 @@ func detectAuthWithRequest(
...
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		logger.Debugf("Failed to check %s: %v", uri, err)
+		return false
+	}
+	defer func() {
+		// Drain and close response body to enable connection reuse
+		_, _ = io.Copy(io.Discard, resp.Body)
```
Comment (amirejaz) [inline pkg/auth/discovery/discovery.go]: It might be good to limit how much of the response body we drain (e.g., up to 1 MB) to avoid unbounded reads and excessive memory use.

**Comment by amirejaz** [comment]
@theJC Thanks for the update! Looks good overall, left a couple of minor comments.

File: pkg/auth/discovery/discovery.go (hunk 2 above)
Comment (theJC) [reply pkg/auth/discovery/discovery.go]: this is refering to how the boolean is determined.
- URI must be accessible
- URI must return json response

I'll swizzle the comment to see if I cant make that more clear

File: pkg/auth/discovery/discovery_test.go [hunk 4]
```diff
@@ -1200,6 +1200,29 @@ func TestCheckWellKnownURIExists_ErrorPaths(t *testing.T) {
...
 		assert.False(t, result, "Expected false for cancelled context")
 	})
+
+	t.Run("large response body is safely drained with limit", func(t *testing.T) {
+		t.Parallel()
+		// Create a server that returns a very large response body
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			// Write 5 MB of data (exceeds MaxResponseBodyDrain of 1 MB)
```
Comment (amirejaz) [inline pkg/auth/discovery/discovery_test.go]: Nice! We can simplify it a bit by writing the response in one shot with bytes.Repeat and reducing it to around 2 MB, that’s still enough to exceed the 1 MB drain cap and will keep the test faster and cleaner.

```
w.Write(bytes.Repeat([]byte("X"), 2*1024*1024))
```

Comment (theJC) [reply pkg/auth/discovery/discovery_test.go]: gotcha.  Done!

File: pkg/auth/discovery/discovery.go (hunk 3 above)
Comment (theJC) [reply pkg/auth/discovery/discovery.go]: good point. addressed

**Comment by amirejaz** [comment]
Thanks a lot for your valuable contribution, @theJC! 🙏
Really appreciate the time and effort you put into refining this, everything looks great now. Once the CI checks pass, we’ll get this merged. 🚀
